src/coverage/coverageService.ts

```diff
@@ -0,0 +1,195 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { sep, join } from "path";
+import * as vscode from "vscode";
+import { appendIdfAndToolsToPath, spawn } from "../utils";
+import { OutputChannel } from "../logger/outputChannel";
+import { Logger } from "../logger/logger";
+import * as idfConf from "../idfConfiguration";
+
+const COVERED_FACTOR = 0.9;
+const PARTIAL_FACTOR = 0.75;
+const MIN_LINE_COUNT = 1;
+
+export interface countRange {
+  range: vscode.Range;
+  count: Number;
+}
+
+export interface textEditorWithCoverage {
+  allLines: vscode.Range[];
+  countPerLines: countRange[];
+  coveredLines: vscode.Range[];
+  editor: vscode.TextEditor;
+  partialLines: vscode.Range[];
+  uncoveredLines: vscode.Range[];
+}
+
+export async function buildJson(dirPath: string) {
+  const idfPathDir =
+    idfConf.readParameter("idf.espIdfPath") || process.env.IDF_PATH;
+  const componentsDir = join(idfPathDir, "components");
+  const result = await _runCmd(
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      componentsDir,
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--json",
+    ],
+    dirPath
+  );
+  return JSON.parse(result);
+}
+
+export async function buildHtml(dirPath: string) {
+  const idfPathDir =
+    idfConf.readParameter("idf.espIdfPath") || process.env.IDF_PATH;
+  const componentsDir = join(idfPathDir, "components");
+  const result = await _runCmd(
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      componentsDir,
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--html",
+    ],
+    dirPath
+  );
+  return result;
+}
+
+function _runCmd(cmd: string, args: string[], dirPath: string) {
+  const modifiedEnv = appendIdfAndToolsToPath();
+  return spawn(cmd, args, { env: modifiedEnv, cwd: dirPath })
+    .then((resultBuffer) => resultBuffer.toString())
+    .catch((e) => {
+      const msg = e.error ? e.error.message : e;
+      Logger.error("Error on gcov cmd.\n" + msg, e);
+      OutputChannel.appendLine("Error building gcov cmd.\n" + msg);
+      return "";
+    });
+}
+
+export async function generateCoverageForEditors(
+  editors: vscode.TextEditor[],
+  dirPath: string
+) {
+  const gcovObj = await buildJson(dirPath);
+  const coveredEditors: textEditorWithCoverage[] = [];
+  for (let editor of editors) {
+    let gcovObjFilePath = "";
+    if (editor.document.fileName.indexOf(dirPath) > -1) {
```

Comment:
This is causing un-expected behaviour; 

### Steps to reproduce:
- Open the `examples/system/gcovr` project 
- Run command `Show editor coverage `
- Put breakpoint on 101, and open `Readme.md` file